source/common/stream_info/stream_info_impl.h

```diff
@@ -218,6 +219,25 @@ struct StreamInfoImpl : public StreamInfo {
     return upstream_transport_failure_reason_;
   }
 
+  void setRequestHeaders(const Http::HeaderMap& headers) override {
+    headers.iterate(
+        [](const Http::HeaderEntry& entry, void* self) {
+          static_cast<StreamInfoImpl*>(self)->request_headers_.emplace_back(
+              std::string(entry.key().getStringView()), entry.value().getStringView());
+          return Http::HeaderMap::Iterate::Continue;
+        },
+        this);
+  }
```

Comment:
Unfortunately we have to copy the headers.

Consider the use-case of renaming a header:

```
              request_headers_to_add:
              - header:
                  key: x-new-name
                  value: "%REQ(x-header)%"
              request_headers_to_remove:
              - x-header
```

it won't work because request_headers_to_remove will do its job before request_headers_to_add.  
Filters can also remove of modify headers.

So there should be a copy.

Maybe we should copy them into the HeaderMapImpl. It would call the same `src_header_map.iterate() ` and copy all the strings. However there is an optimisation TODO in HeaderMapImpl::copyFrom , and indeed it's possible to avoid certain copies (header names?)
Also we could try the HeaderString::setReference, but that'd be very dangerous as it's hard to prove all the lifetimes.

Reply:
My opinion is that copying all the incoming headers on every request is not acceptable from a performance perspective. But even if the copy could be skipped unless needed, I think users of this feature will expect that headers modifications made by other filters (including the router itself) will be visible to this feature.

Renaming headers is a valid use case, but I think we need to come up with a different way to handle it.

Reply:
I agree that it's not good to copy all the headers even when this `%REQ()%` feature is not used anywhere in the config.

Another concern is that the `%REQ()%` in access logs behaves differently as it uses a reference to the request headers, not a copy of the original headers.

So yeah it would be better to use a reference here.

As for renaming headers, we could add a `repeated string preserve_request_headers` config section, and add a `%O[riginal]REQ(header-name)%` header formatter. But that'd be a different PR/issue.